Add an initial testsuite

obnoxxx opened the pull request:
This is based on the original ideas of @MohamedAshiqrh 's PR #184 .
I basically took (and slightly) extended the kinds of commands that are run,
but put it into more of a test framework. (Copied the tiny "subunit.sh" from samba).
So now it really fails when a test fails. Also integrated this into "make test" and into .travis.yml.

I am pushing this to get some comments.
It currently contains the patches of PR #193 as well.
If we agree that this is the way to go, and if we merge #193, then I'd rebase once more to make git log clean.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gluster/gluster-kubernetes/194)
<!-- Reviewable:end -->

obnoxxx commented:
ping @MohamedAshiqrh @jarrpa @raghavendra-talur @humblec @lpabon 

what do you think? Can we merge this (possibly after merging #193 and rebasing, but not necessarily) to get started with some (hopefully increasing) test coverage?

obnoxxx commented:
ping, any thoughts? :-)

At some point I will start asking: any objections? ;-)

MohamedAshiqrh commented:
Add a README for prerequisite.

yamllint - This is the only one now.

MohamedAshiqrh commented:
We can also maintain README to say the TO-DO's

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+
+failed=0
+
+testit "test missing toplogy" \
```
Spelling

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+
+test_cli_not_found () {
+	local saved_path="${PATH}"
+	PATH="/usr/bin:/bin"
```
oc and kube usually stay in this path. If they are properly installed then the test will not pass right ? 

Does this mean We can not run test on machine which has the oc and kube installed?

Am I missing something?

obnoxxx replied:
That's right. Going to use a different path that won't work, maybe even does not exist. Like "/foo".

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+	OUT=$(${GK_DEPLOY} -y ${TOPOLOGY})
+	local rc=${?}
+
+	if [ "x${rc}" = "x0" ]; then
```
==

obnoxxx replied:
This is a bash-ism. While we are using bash,
i still like to keep the bash-isms at a minimum and use posix-compatible alternatives where it does not disturb. similar for [[ vs [ . If more people want it the bash way, I can do that...

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+	OUT=$(${GK_DEPLOY} -y -c ${cli} ${TOPOLOGY})
+	local rc=$?
+
+	if [ "x${rc}" = "x0" ]; then
```
==

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+
+	if [ "x${#}" != "x0" ]; then
+		cli="${1}"
+		if [ "x${cli}" = "xkubectl" ]; then
```
==

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+			expected_out="Using Kubernetes CLI.
+Namespace 'invalid' not found."
+		elif [ "x${cli}" != "xoc" ]; then
+			expected_out="Container platform CLI (e.g. kubectl, oc) not found."
```
expected out put is 

"Unknown CLI '${cli}'."

obnoxxx replied:
hm. no.... the test succeds?

MohamedAshiqrh replied:
I tried "test_namespace_invalid foo"
it said test failed. But this block is supposed to handle if someone enters neither os nor kubeclt. This fails

obnoxxx replied:
done. good catch! (added corresponding test run)

MohamedAshiqrh commented on tests/yaml/test_syntax.sh:
```diff
@@ -0,0 +1,49 @@
+OGTY="${OTD}/glusterfs-template.yaml"
+OHTY="${OTD}/heketi-template.yaml"
+OHSY="${OTD}/heketi-service-account.yaml"
+
```
deploy-heketi missing. because this was done before revert of heketi 4.

obnoxxx replied:
done

MohamedAshiqrh commented on tests/yaml/test_syntax.sh:
```diff
@@ -0,0 +1,49 @@
+	check_invalid_yaml ${FAULTY_YAML} \
+	|| failed=$((failed + 1))
+
+for yaml in ${KGDY} ${KHDY} ${KHSY} ${OGTY} ${OHTY} ${OHSY} ; do
```
deploy heketi missing

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,157 @@
+GK_DEPLOY="${DEPLOY_DIR}/gk-deploy"
+TOPOLOGY="${DEPLOY_DIR}/topology.json.sample"
+
+PATH="${STUBS_DIR}:$PATH"
```
Cool.

But is it removed from path somewhere?

MohamedAshiqrh replied:
I mean cleaned up? Interestingly my PATH looks clean without stub and I don't see a place where it is cleaned up.

obnoxxx replied:
You don't need to clean that up. PATH will only be modified for the script. When returning to the caller, the path is as before. That is because executing the script runs it in a new shell. Only if you source the script instead of executing it, you will have the caller's path changed.

obnoxxx replied:
done

MohamedAshiqrh commented on tests/yaml/test_syntax.sh:
```diff
@@ -0,0 +1,49 @@
+
+check_invalid_yaml () {
+	check_yaml ${1}
+	if [ "x$?" = "x0" ]; then
```
==

MohamedAshiqrh commented on tests/run.sh:
```diff
@@ -0,0 +1,22 @@
+		continue
+	fi
+
+	pushd ${testdir}
```
nice

obnoxxx commented:
rebased and added patches on top to address (most of) @MohamedAshiqrh 's requests.

obnoxxx commented:
@MohamedAshiqrh you are right, good catch! This case was not caught in the test runs. added a test run and fixed the test func. will squash the SQ patches later . leaving them on top for now so you can see better what's going on.

MohamedAshiqrh commented:
@obnoxxx looks good 👍

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -97,7 +97,7 @@ Namespace 'invalid' not found."
 Namespace 'invalid' not found."
 		elif [ "x${cli}" != "xoc" ]; then
-			expected_out="Container platform CLI (e.g. kubectl, oc) not found."
+			expected_out="Unknown CLI '${cli}'."
```
cool

obnoxxx commented:
@MohamedAshiqrh I rebased once more and:

* Changed the code to use  == and also [[ ... ]] instead of [ ... ] in the checks, since this anyways the style in gk-deploy.
* Added a tests/README with prereqs and todos.

According to your previous comments, this should resolve the remaining points.

Will squash (the to-be-squashed patches) and merge when you're ok.

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,161 @@
+
+test_cli_not_found () {
+	local saved_path="${PATH}"
+	PATH="/doesnotexist"
```
👍

MohamedAshiqrh commented on tests/README.md:
```diff
@@ -0,0 +1,19 @@
+* More elaborate basic tests need fuller mocking/stubbing of tools
+* Write full functional tests to be run in vms
+ (like the kubernetes vagrant environment)
+
```
👍 This looks nice.

MohamedAshiqrh commented on tests/gk-deploy/stubs/kubectl:
```diff
@@ -0,0 +1,20 @@
+
+	NAMESPACE="$3"
+
+	if [ "$NAMESPACE" = "invalid" ]; then
```
Missed here == and [[ ]]

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/stubs/kubectl:
```diff
@@ -0,0 +1,20 @@
+
+SCRIPT_DIR=$(cd $(dirname $0) ; pwd)
+
+if [ "x$1" = "xget" ]; then
```
Missed here == and [[ ]]

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/stubs/kubectl:
```diff
@@ -0,0 +1,20 @@
+	if [ "$NAMESPACE" = "invalid" ]; then
+		exit 1
+	fi
+elif [ "x$1" = "xconfig" ]; then
```
here

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/stubs/oc:
```diff
@@ -0,0 +1,20 @@
+
+SCRIPT_DIR=$(cd $(dirname $0) ; pwd)
+
+if [ "x$1" = "xget" ]; then
```
here

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/stubs/oc:
```diff
@@ -0,0 +1,20 @@
+
+	NAMESPACE="$3"
+
+	if [ "$NAMESPACE" = "invalid" ]; then
```
here

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/stubs/oc:
```diff
@@ -0,0 +1,20 @@
+	if [ "$NAMESPACE" = "invalid" ]; then
+		exit 1
+	fi
+elif [ "x$1" = "xconfig" ]; then
```
here

obnoxxx replied:
done

MohamedAshiqrh commented on tests/gk-deploy/test_gk_deploy_basic.sh:
```diff
@@ -0,0 +1,161 @@
+	test_namespace_invalid kubectl \
+	|| failed=$((failed + 1))
+
+testit "test namespace invalid unknown-cli" \
```
nice

MohamedAshiqrh commented:
@obnoxxx I know kind of annoying "here", bare with me this time :-) . Thanks for doing this, I am looking forward to add more tests to it.

MohamedAshiqrh commented:
@obnoxxx This make me happy. No more bad yamls.

```

time: 2017-03-24 13:40:49.989625Z
failure: check glusterfs-daemonset.yaml [
/home/ashiq/Work/gluster-kubernetes/tests/yaml/../../deploy/kube-templates/glusterfs-daemonset.yaml:67:8: [error] syntax error: expected <block end>, but found '<block mapping start>'
]
time: 2017-03-24 13:40:49.992013Z
test: check heketi-deployment.yaml
time: 2017-03-24 13:40:50.174580Z
success: check heketi-deployment.yaml

```

obnoxxx commented:
@MohamedAshiqrh not annoying. This is exactly the kind of thorough reviews we need and I want... :-)

obnoxxx commented:
will update later with changes . i just had forgotten the stub files.

obnoxxx commented:
fixed the mentioned places.
also fixed tests/run.sh
did NOT fix subunit.sh since this was copied from outside.

I guess it should be all fine now?

MohamedAshiqrh commented:
@obnoxxx All looks good. Really nice work.

obnoxxx commented:
Thanks @MohamedAshiqrh, merging...

obnoxxx commented:
Ooooh gosh, I merged the PR without first squashing the SQ commits into their respecitive actual commits (they were supposed to fix)... :-( 

And I only wanted to be nice and hence put the SQ commits separately on top. Now I know again why usually want to amend the patches right away ... hence you do not forget...

Sorry for the messed up git history!!!  :-(
